python/pyspark/sql/group.py

```diff
@@ -194,6 +194,37 @@ def pivot(self, pivot_col, values=None):
             jgd = self._jgd.pivot(pivot_col, values)
         return GroupedData(jgd, self.sql_ctx)
 
+    def apply(self, udf_obj):
+        """
+        Maps each group of the current [[DataFrame]] using a pandas udf and returns the result
+        as a :class:`DataFrame`.
+
+        """
+        from pyspark.sql.functions import pandas_udf
+
+        if not udf_obj._vectorized:
```

Comment:
You should first check to make sure it is an instance of a `UserDefinedFunction`.

Reply:
Yeah.. the udf_obj here is actually a `<class 'function'>` because `udf` and `pandas_udf` returns a `UserDefinedFunction.wrapped()` (to attach doc string I believe).

Maybe it possible to change `udf` and `pandas_udf` to return a `UserDefinedFunction` object and attach doc string to its `__call__` methods. I can give it a try.

Reply:
It seems `foo?` and `foo??` only shows the property doc string if `foo` is a `function`.

If `foo` is a `UserDefinedFunction`, it will show the docstring for the class.

For that reason, I think we should keep the return value of `udf` and `pandas_udf` to be a wrapped function. And I will do checks on `udf_obj.func` to see if this is a valid wrapped udf function. @BryanCutler what do you think?

Reply:
I ended up checking `hasattr(input, 'func')` to check if it's a valid input. It's not great but I don't what's better though.

Reply:
You can just check if it has that attribute first then `if not hasattr(udf_obj, "_vectorized") or not udf_obj._vectorized`

Reply:
Fixed.